### PR TITLE
feat: make Postgres connection pool settings configurable (pgBouncer transaction-mode / NullPool support)  

### PR DESCRIPTION
**Pull request by pythonmonty**
## Summary & Motivation

Implements #33935.

Make the Postgres storage SQLAlchemy connection pool configurable, and add a `pool_mode: transaction` option that switches the engine to a `NullPool` so Dagster's Postgres storage can run behind pgBouncer in transaction-pooling
mode.

Today, when `dagster-webserver` / `dagster dev` start, each storage rebuilds its engine in `optimize_for_webserver()` with a hardcoded `pool_size=1`, and only `pool_recycle` / `max_overflow` are tunable (webserver CLI flags only). There is no way to:
- set `pool_size`,
- configure any pool setting via `dagster.yaml` (so daemon / code-server processes have no available configuration), or
- opt out of persistent pooled connections, which is required for pgBouncer transaction pooling (a `QueuePool` holding sessions open is incompatible with transaction-level pooling).

This PR exposes the pool parameters as first-class config and threads them through the storages, CLIs, and Helm chart.

### Changes

- **`dagster.yaml` storage config** (`pg_config()` / `PostgresStorageConfig`): new optional `pool_size`, `max_overflow`, `pool_recycle`, `pool_mode` fields, read in `from_config_value()` and applied in `optimize_for_webserver()`.
  - `pool_mode == "transaction"` -> with `poolclass=NullPool` (no persistent connections; pgBouncer transaction-mode compatible).
  - otherwise `QueuePool` with the configured values, each falling back to the CLI-provided value when unset (config takes precedence over CLI flags).
- **CLI**: new `--db-pool-size` on `dagster-webserver`, `dagster dev`, and `dg dev`, threaded through `DagsterInstance.optimize_for_webserver()`. Defaults to `1`, preserving current behavior.
- **Helm chart**: `dagsterWebserver.dbPoolSize` maps to `--db-pool-size`, and `postgresql.pool.{size,maxOverflow,recycle,mode}` written into `dagster.yaml` (with `values.schema.json` updated).
- **MySQL storages**: `optimize_for_webserver` signature updated to accept the new `pool_size` param (default `1`) so the shared interface stays consistent. Required because `DagsterInstance.optimize_for_webserver` now forwards `pool_size` to every storage. Without this, MySQL deployments would raise `TypeError` at webserver startup (NullPool/pgBouncer semantics stay Postgres-specific.)

Backwards compatibility: defaults reproduce the previous behavior (`pool_size=1`, existing recycle/overflow defaults), so deployments that don't set any of the new options are unaffected.

**Note on `pool_size=0`** for SQLAlchemy `QueuePool`, `pool_size=0` means *no limit* (unlimited persistent connections), not disabled. To drop persistent connections for pgBouncer transaction mode, use `pool_mode: transaction` (aka `NullPool`), not `pool_size: 0`. CLI/Helm help text states this explicitly.

### Related issues

- This PR addresses the pooling/connection-management side of #33416 (pgbouncer transaction mode compatibility)
by adding the `NullPool` path. It does not by itself resolve the separate concern in that issue that statement timeouts are applied via a bare `SET statement_timeout = ...` (`set_pg_statement_timeout`), which pgBouncer transaction mode does not support (`SET LOCAL` would be needed). This change will be a follow-up PR.
- This PR also gives user a configuration relevant to #20325 and #8466.

## Test Plan

New Helm schema tests:

- `test_storage_postgres_pool_config` asserts `pool_size`/`max_overflow`/ `pool_recycle` (including the `pool_size=0` edge case) render into `dagster.yaml` for run / schedule / event-log storage.
- `test_storage_postgres_pool_mode` asserts `pool_mode: transaction` renders.
- `test_webserver_db_pool_size` asserts `--db-pool-size` is passed to the webserver command (incl. `0`).

`pytest python_modules/libraries/dagster-postgres/dagster_postgres_tests` and `.../dagster-mysql/dagster_mysql_tests` (verify `optimize_for_webserver` still works through `DagsterInstance` for both backends after the signature change).

Running `make ruff` locally produced a clean result, no changes needed.

Tests I ran locally:

| Test suite                                                                  | Result |
|-----------------------------------------------------------------------------|--------|
| `helm/dagster/schema/schema_tests/test_dagit.py` + `test_instance.py`       | passed |
| `python_modules/dagster/dagster_tests/core_tests/instance_tests/`           | passed |
| `dagster-postgres/dagster_postgres_tests/test_instance.py::test_conn_str`   | passed |
| Core storage tests (keyword `storage`)                                      | passed |

The DB-dependent postgres tests (`test_statement_timeouts`, `test_connection_leak`, etc) require a running PostgreSQL server and were skipped. All locally runnable tests for the pool mode feature have passed.

### CODE REVIEW

File: python_modules/dagster-webserver/dagster_webserver/cli.py [hunk 1]
```diff
@@ -132,6 +133,17 @@ def create_dagster_webserver_cli():
...
+    "--db-pool-size",
+    help=(
+        "The number of connections to keep open in the sqlalchemy pool. Set to 0 to disable "
+        "persistent connections, which is required for pgBouncer transaction-mode pooling. "
+        "Not respected in all configurations."
+    ),
+    default=DEFAULT_POOL_SIZE,
+    type=click.INT,
+    show_default=True,
+)
```
[the remark this answers is not in the document]
Comment (pythonmonty) [reply python_modules/dagster-webserver/dagster_webserver/cli.py:147]: Adapted already in [this commit](https://github.com/dagster-io/dagster/pull/33936/changes/c1049a4c3bc37af518e6bba79f43929f2a02e366)

File: python_modules/dagster/dagster/_cli/dev.py [hunk 2]
```diff
@@ -129,6 +129,15 @@
...
 )
+@click.option(
+    "--db-pool-size",
+    help=(
+        "The number of connections to keep open in the sqlalchemy pool. Set to 0 to disable "
+        "persistent connections, which is required for pgBouncer transaction-mode pooling."
+    ),
+    default=None,
+    type=click.INT,
+)
```
[the remark this answers is not in the document]
Comment (pythonmonty) [reply python_modules/dagster/dagster/_cli/dev.py:142]: Adapted already in [this commit](https://github.com/dagster-io/dagster/pull/33936/changes/c1049a4c3bc37af518e6bba79f43929f2a02e366)

File: python_modules/libraries/dagster-dg-cli/dagster_dg_cli/cli/dev.py [hunk 3]
```diff
@@ -86,6 +86,15 @@
...
 )
+@click.option(
+    "--db-pool-size",
+    help=(
+        "The number of connections to keep open in the sqlalchemy pool. Set to 0 to disable "
+        "persistent connections, which is required for pgBouncer transaction-mode pooling."
+    ),
+    default=None,
+    type=click.INT,
+)
```
[the remark this answers is not in the document]
Comment (pythonmonty) [reply python_modules/libraries/dagster-dg-cli/dagster_dg_cli/cli/dev.py:99]: Adapted already in [this commit](https://github.com/dagster-io/dagster/pull/33936/changes/c1049a4c3bc37af518e6bba79f43929f2a02e366)

**Comment by pythonmonty** [comment]
Could someone from the maintainers review please?

**Comment by daankaiko** [comment]
Also really waiting for this feature to be merged.

**Comment by darcywalsh-vq** [comment]
Also keen for this to be available

**Comment by TabLand** [comment]
~[raayu83](https://github.com/raayu83) , [cmpadden](https://github.com/cmpadden), [gibsondan](https://github.com/gibsondan)~
@prha  / @smackesey

**Comment by natalie-kaiko** [comment]
Really looking forward to this being available! 🙏

**Comment by jan-benisek-lgtm** [comment]
That's also something we'd benefit from a lot! Could this be merged?
